File: prepa/tata.js

// node tata.js

const fs = require('fs');

function parseInputs(file_path) {
    const data = fs.readFileSync(file_path, 'utf8');
    return data.split(/\r?\n/)
}

function getPriority(c) {
    if(c > 'a' && c <= 'z') {
        return 1+c.charCodeAt(0)-'a'.charCodeAt(0)
    } else if(c >= 'A' && c <= 'Z') {
        return 27+c.charCodeAt(0)-'A'.charCodeAt(0)
    } else {
        return 0 // Oups
    }
}

function exo1(inputs) {
    priority = 0
    inputs.forEach(line => {
        compartment_1 = line.substring(0, line.length/2)
        compartment_2 = line.substring(line.length/2, line.length)
        compartment_1 = [...compartment_1].reduce((acc, curr)=> {
            return acc.includes(curr) ? acc : acc + curr;
        }, "")
        compartment_1.split("").forEach(c => {
            if( compartment_2.includes(c) ){
                priority += getPriority(c)
            }
        })
    })
    return priority
}

function exo2(inputs) {
    priority = 0
    for (let i = 0; i < inputs.length/3; i++) { 
        blah = [...inputs[i*3]].reduce((acc, curr)=> {
            return acc.includes(curr) ? acc : acc + curr;
        }, "")
        blah.split("").forEach(c => {
            if( inputs[i*3+1].includes(c) && inputs[i*3+2].includes(c) ){
                priority += getPriority(c)
            }
        })
    }
    return priority
}


inputs = parseInputs("input.txt")

answ1 = exo1(inputs)
console.log("The first answer is "+answ1)
answ2 = exo2(inputs)
console.log("The second answer is "+answ2)

File: prepa/tata.java

// java tata.java

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.ArrayList;

class Tata {
    
    public static List<String> parseInputs(String filepath) throws IOException {
        return Files.readAllLines(Paths.get(filepath));
    }
    
    public static int getPriority(char c) {
        if(c >= 'a' && c <= 'z') {
            return 1+c-'a';
        } else if( c>='A' && c<='Z') {
            return 27+c-'A';
        } else {
            return 0; // HOHO            
        }
    }
    
    public static int exo1(List<String> inputs) {
        int priority = 0;
        for(String rucksack : inputs) {
            String compartment_1 = rucksack.substring(0, rucksack.length()/2);
            String compartment_2 = rucksack.substring(rucksack.length()/2, rucksack.length());
            List<Character> found_char = new ArrayList<>();
            for(char c : compartment_1.toCharArray()) {
                if( (found_char.contains(c) == false) && (compartment_2.indexOf(c) > -1)) {
                    priority += getPriority(c);
                    found_char.add(c);
                }
            }
        }
        return priority;
    }

    public static int exo2(List<String> inputs) {
        int priority = 0;
        for(int i = 0; i < inputs.size()/3; ++i) {
            List<Character> found_char = new ArrayList<>();
            for(char c : inputs.get(3*i).toCharArray()) {
                if( (found_char.contains(c) == false) && (inputs.get(3*i+1).indexOf(c) > -1) && (inputs.get(3*i+2).indexOf(c) > -1))
                {
                    priority += getPriority(c);
                    found_char.add(c);
                }
            }
        }
        return priority;
    }
    
    public static void main(String[] args) {
        try {
            List<String> inputs = parseInputs("input.txt");
            int answ1 = exo1(inputs);
            System.out.println("The first answer is "+answ1); 
            int answ2 = exo2(inputs);
            System.out.println("The second answer is "+answ2); 
            
        } catch (IOException e) {
			e.printStackTrace();
		}
    }
}